File: 1356-sort-integers-by-the-number-of-1-bits/1356-sort-integers-by-the-number-of-1-bits.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */
var sortByBits = function(arr) {
  
    
    return arr.sort((a,b) => {
        const bA = numBits(a);
        const bB = numBits(b);
        
        const d = bA - bB;
        
        if (d!==0) return d
        else return a-b
    })
};


function numBits(n){
    if (n===0) return 0;
    
  
    
    const h = Math.floor(n/2);

    if ( (n % 2) === 1) {
        const b = numBits(h) + 1;
      
        return b
    }
    
    const b = numBits(h);
    
    return b
    
};

File: 1356-sort-integers-by-the-number-of-1-bits/1356-sort-integers-by-the-number-of-1-bits.go
func sortByBits(arr []int) []int {
   
    sort.Slice(arr,func(i, j int) bool {
        
        a, b := nBits(arr[i]), nBits(arr[j]);
        
        switch {
            case a==b:
                return arr[i]<arr[j]
            default:
                return a<b
            
        }
	})
    
    return arr
}

func nBits(n int) int {
    r := 0;
    
    for (n > 0) {
        r += n & 1;
        n = n >> 1;
    }
    return r
}